src/session_pool/session.js

```diff
@@ -1,14 +1,52 @@
 import { cryptoRandomObjectId } from 'apify-shared/utilities';
 import log from 'apify-shared/log';
 import { checkParamOrThrow } from 'apify-client/build/utils';
-import tough from 'tough-cookie';
-import EVENTS from './events';
+import { Cookie, CookieJar } from 'tough-cookie';
+import EVENTS from './events'; // eslint-disable-line import/named,no-unused-vars
 import { STATUS_CODES_BLOCKED } from '../constants';
 import { getCookiesFromResponse } from './session_utils';
 
-const { Cookie } = tough;
+// TYPE IMPORTS
+/* eslint-disable no-unused-vars,import/named,import/no-duplicates,import/order */
+import { EventEmitter } from 'events';
+import { Cookie as PuppeteerCookie } from 'puppeteer';
+/* eslint-enable no-unused-vars,import/named,import/no-duplicates,import/order */
 
+/**
+ * Persistable {Session} state.
+ * @typedef {Object} SessionState
+ * @property {String} id
+ * @property {Object} cookies
+ * @property {Object} cookieJar
+ * @property {Object} userData
+ * @property {Number} maxErrorScore
+ * @property {Number} errorScoreDecrement
+ * @property {String} expiresAt
+ * @property {String} createdAt
+ * @property {Number} usageCount
+ * @property {Number} errorScore
+ */
 
+/**
+ * @typedef {Object} SessionOptions
+ * @property {String} [id] - Id of session used for generating fingerprints. It is used as proxy session name.
+ * @property {Number} [maxAgeSecs=3000] - Number of seconds after which the session is considered as expired.
+ * @property {Object} userData - Object where custom user data can be stored. For example custom headers.
+ * @property {number} [maxErrorScore=3] - Maximum number of marking session as blocked usage.
+ *   If the `errorScore` reaches the `maxErrorScore` session is marked as block and it is thrown away.
+ *   It starts at 0. Calling the `markBad` function increases the `errorScore` by 1.
+ *   Calling the `markGood` will decrease the `errorScore` by `errorScoreDecrement`
+ * @property {number} [errorScoreDecrement=0.5] - It is used for healing the session.
+ *   For example: if your session is marked bad two times, but it is successful on the third attempt it's errorScore is decremented by this
+ *   number.
+ * @property {Date} [createdAt] - Date of creation.
+ * @property {Date} [expiresAt] - Date of expiration.
+ * @property {Number} [usageCount=0] - Indicates how many times the session has been used.
+ * @property {Number} [errorCount=0] - Indicates how many times the session is marked bad.
+ * @property {Number} [maxUsageCount=50] - Session should be used only a limited amount of times.
+ *   This number indicates how many times the session is going to be used, before it is thrown away.
+ * @property {EventEmitter} sessionPool - SessionPool instance. Session will emit the `sessionRetired` event on this instance.
```

Comment:
should be `SessionPool`. the generated typings are coming as `import { EventEmitter } from "node/events";` on `types/session_pool/session.d.ts`

Reply:
tl;dr

That would create a cyclic-dependency, the build won't be happy with this. I am not sure if such a cycle is bad or marginally acceptable. If @mnmkng is OK with a cycle, I am OK as well. 

----

There is a way to compare different solutions to dependency problems - the principles of _cohesion_ and _coupling_. `SessionPool` inherits from `EventEmitter`. This is one possible attempt to achieve high-cohesion and low-coupling across files.

Looking at it from the _Java perspective_, a cycle is OK, as `session/*` can be viewed on as a _package_ - it is acceptable to have higher coupling across a single _package_ as long as high cohesion is maintained across dependent _packages_.

But a _Javaist_ would probably create interfaces for everything and conceptually they can depend on each other and conceptually increasing cohesion while maintaining low coupling.  If forced to avoid cycles, he would put `ISessionPool` and `ISession` into a single file, creating another cycle, but now spanning just the single file :P

Reply:
yeah, I agree, but circular dependency is a non-issue when importing for types. but the right code is indeed `EventEmitter`, since you can "listen" to the event on any `EventEmitter` instance, which `SessionPool` happens to listen on RETIRED and only that. 
using the "interface" of EventEmitter is enough (that is, that at least implements `emit`, can even be an object `{ emit: () => }`)

Reply:
Technically, cycles are fine, unless someone adds something that breaks stuff. Can we use some typescript magic to get rid of the cycle? I don't know, strip type imports at build time?

Reply:
there's a new TS-only keyword that only works on types and not on concrete imports, but the file must be .ts

```ts
import type { EventEmitter } from 'events';
```

the generated code has no mention to 'events' at all, only the declarations. maybe babel will add those as a plugin/transform to catch-up. https://github.com/babel/babel/issues/10981

Reply:
Done.